src/core/stdcpp/string.d

```diff
@@ -0,0 +1,428 @@
+                _Getal() = al;
+        }
+
+        void _Tidy_init() nothrow @nogc
+        {
+            auto _My_data = &_Get_data();
+            _My_data._Mysize = 0;
+            _My_data._Myres = _My_data._BUF_SIZE - 1;
+            _My_data._Bx._Buf[0] = T(0);
+        }
+
+        size_type _Calculate_growth(size_type _Requested) const nothrow @nogc
+        {
+            auto _My_data = &_Get_data();
+            size_type _Masked = _Requested | _My_data._ALLOC_MASK;
+            size_type _Old = _My_data._Myres;
+            size_type _Expanded = _Old + _Old / 2;
+            return _Masked > _Expanded ? _Masked : _Expanded;
+        }
+
+        ref basic_string _Reallocate_for(_ArgTys...)(size_type _New_size, void function(pointer, size_type, _ArgTys) nothrow @nogc _Fn, _ArgTys _Args) nothrow @nogc
+        {
+            auto _My_data = &_Get_data();
+            size_type _Old_capacity = _My_data._Myres;
+            size_type _New_capacity = _Calculate_growth(_New_size);
+            auto _Al = &_Getal();
+            pointer _New_ptr = _Al.allocate(_New_capacity + 1); // throws
+            _Base._Orphan_all();
+            _My_data._Mysize = _New_size;
+            _My_data._Myres = _New_capacity;
+            _Fn(_New_ptr, _New_size, _Args);
+            if (_My_data._BUF_SIZE <= _Old_capacity)
+                _Al.deallocate(_My_data._Bx._Ptr, _Old_capacity + 1);
+            _My_data._Bx._Ptr = _New_ptr;
+            return this;
+        }
+
+        ref basic_string _Reallocate_grow_by(_ArgTys...)(size_type _Size_increase, void function(pointer, const(T)[], _ArgTys) nothrow @nogc _Fn, _ArgTys _Args) nothrow @nogc
+        {
+            auto _My_data = &_Get_data();
+            size_type _Old_size = _My_data._Mysize;
+            size_type _New_size = _Old_size + _Size_increase;
+            size_type _Old_capacity = _My_data._Myres;
+            size_type _New_capacity = _Calculate_growth(_New_size);
+            auto _Al = &_Getal();
+            pointer _New_ptr = _Al.allocate(_New_capacity + 1); // throws
+            _Base._Orphan_all();
+            _My_data._Mysize = _New_size;
+            _My_data._Myres = _New_capacity;
+            if (_My_data._BUF_SIZE <= _Old_capacity)
+            {
+                pointer _Old_ptr = _My_data._Bx._Ptr;
+                _Fn(_New_ptr, _Old_ptr[0 .. _Old_size], _Args);
+                _Al.deallocate(_Old_ptr, _Old_capacity + 1);
+            }
+            else
+                _Fn(_New_ptr, _My_data._Bx._Buf[0 .. _Old_size], _Args);
+            _My_data._Bx._Ptr = _New_ptr;
+            return this;
+        }
+
+        static void _Xran() @trusted @nogc { _Xout_of_range("invalid string position"); }
+        static void _Xlen() @trusted @nogc { _Xlength_error("string too long"); }
+
+        _String_alloc!(_String_base_types!(T, Alloc)) _Base;
+    }
+    else
+    {
+        static assert(false, "C++ runtime not supported");
```

Comment:
@TurkeyMan  you don't have to actually implement the `extern(C++)` functions, a declaration should suffice. In fact only functions that you want inlinable by DMD (that excludes platforms that are not of interest to you, if someone else wants that performance, let them do it) need implementations. LTO will solve perf problems on other compilers.
Things that are `extern(D)` e.g. constructors that assign `const(T)[]` need to be implemented in D but can be done later as they are a convenience. 

What _is_ required are the data members. Doesn't matter what the layout is if you never use them, anything that adds up to 12 (for 32bit0 or 24 (for 64 bit) bytes will do provided it is aligned to 4 or 8 bytes.

Reply:
I thought that at first. But reality is, link errors errors are inevitable. I've come to the conclusion that only hard-functions can extern (allocation, etc). Template functions must mirror. And that's fine... They're pretty simple really.

Reply:
Ah, I forgot about template functions. 

> But reality is, link errors errors are inevitable

Ouch, I thought the whole purpose of `extern(C++)` was link compatibility.

> They're pretty simple really.

Good luck then ;)

Reply:
Basically, user experience is terrible unless I write at least the functions you see here.
They match the C++ runtimes exactly... but they don't need to. I could consolidate them if we didn't care about having a perfect match.
It's really just the allocator functions that have to work right, and the data structures.

Reply:
Fair enough. Good luck!

Reply:
Luck indeed! The hardest part is getting tests to work... And the brutal agonising level of patience required to move anything forward here ;)